File: libs/sdk-typescript/src/utils/otel.decorator.ts
/*
 * Copyright 2025 Daytona Platforms Inc.
 * SPDX-License-Identifier: Apache-2.0
 */

import { trace, context as otelContext, metrics, SpanStatusCode, Histogram } from '@opentelemetry/api'

// Lazy initialization to ensure SDK is started before getting tracer/meter
const getTracer = () => trace.getTracer('')
const getMeter = () => metrics.getMeter('')

const executionHistograms = new Map<string, Histogram>()

/**
 * Configuration options for span instrumentation
 */
export interface SpanConfig {
  /**
   * Custom name for the span. If not provided, uses `ClassName.methodName` format
   */
  name?: string
  /**
   * Additional attributes to attach to the span
   */
  attributes?: Record<string, string>
}

/**
 * Configuration options for metric instrumentation
 */
export interface MetricConfig {
  /**
   * Custom name for the metric. If not provided, uses `ClassName.methodName` format
   */
  name?: string
  /**
   * Description for the metrics being collected
   */
  description?: string
  /**
   * Additional labels to attach to the metrics
   */
  labels?: Record<string, string>
}

/**
 * Configuration options for the combined instrumentation decorator
 */
export interface InstrumentationConfig {
  /**
   * Custom name for the span and metric. If not provided, uses `ClassName.methodName` format
   */
  name?: string
  /**
   * Description for the metrics being collected
   */
  description?: string
  /**
   * Additional labels/attributes to attach to spans and metrics
   */
  labels?: Record<string, string>
  /**
   * Enable trace collection (default: true)
   */
  enableTraces?: boolean
  /**
   * Enable metrics collection (default: true)
   */
  enableMetrics?: boolean
}

type AsyncMethod<This, Args extends any[], Return> = (this: This, ...args: Args) => Promise<Return>

/**
 * Converts a string to snake_case for Prometheus-friendly metric names
 */
function toSnakeCase(str: string): string {
  return str
    .replace(/([A-Z])/g, '_$1')
    .toLowerCase()
    .replace(/^_/, '')
    .replace(/\./g, '_')
}

/**
 * Decorator for instrumenting methods with OpenTelemetry spans (traces only)
 *
 * @param config - Configuration object or string name for the span
 */
export function WithSpan(config?: string | SpanConfig) {
  return function <This, Args extends any[], Return>(
    originalMethod: AsyncMethod<This, Args, Return>,
    context: ClassMethodDecoratorContext<This, AsyncMethod<This, Args, Return>>,
  ): AsyncMethod<This, Args, Return> {
    const methodName = String(context.name)

    return async function (this: This, ...args: Args): Promise<Return> {
      const cfg: SpanConfig = typeof config === 'string' ? { name: config } : config || {}
      const { name, attributes = {} } = cfg

      const className = (this as any)?.constructor?.name ?? 'Anonymous'
      const spanName = name || `${className}.${methodName}`

      const allAttributes = {
        component: className,
        method: methodName,
        ...attributes,
      }

      const span = getTracer().startSpan(spanName, { attributes: allAttributes }, otelContext.active())

      return otelContext.with(trace.setSpan(otelContext.active(), span), async () => {
        try {
          const result = await originalMethod.apply(this, args)
          span.setStatus({ code: SpanStatusCode.OK })
          return result
        } catch (error) {
          span.setStatus({
            code: SpanStatusCode.ERROR,
            message: error instanceof Error ? error.message : String(error),
          })
          span.recordException(error instanceof Error ? error : new Error(String(error)))
          throw error
        } finally {
          span.end()
        }
      })
    }
  }
}

/**
 * Decorator for instrumenting methods with OpenTelemetry metrics (metrics only)
 *
 * Collects:
 * - Histogram: `{name}_duration` - tracks execution duration in milliseconds, with a `status` label (success/error)
 *
 * @param config - Configuration object or string name for the metric
 */
export function WithMetric(config?: string | MetricConfig) {
  return function <This, Args extends any[], Return>(
    originalMethod: AsyncMethod<This, Args, Return>,
    context: ClassMethodDecoratorContext<This, AsyncMethod<This, Args, Return>>,
  ): AsyncMethod<This, Args, Return> {
    const methodName = String(context.name)

    return async function (this: This, ...args: Args): Promise<Return> {
      const cfg: MetricConfig = typeof config === 'string' ? { name: config } : config || {}
      const { name, description, labels = {} } = cfg

      const className = (this as any)?.constructor?.name ?? 'Anonymous'
      const metricName = toSnakeCase(name || `${className}.${methodName}`)
      const allLabels = {
        component: className,
        method: methodName,
        ...labels,
      }

      if (!executionHistograms.has(metricName)) {
        executionHistograms.set(
          metricName,
          getMeter().createHistogram(`${metricName}_duration`, {
            description: description || `Duration of executions for ${metricName}`,
            unit: 'ms',
          }),
        )
      }
      const histogram = executionHistograms.get(metricName)
      if (!histogram) {
        throw new Error(`Histogram not found for metric: ${metricName}`)
      }

      const startTime = Date.now()
      let status: 'success' | 'error' = 'success'
      try {
        const result = await originalMethod.apply(this, args)
        return result
      } catch (error) {
        status = 'error'
        throw error
      } finally {
        const duration = Date.now() - startTime
        histogram.record(duration, { ...allLabels, status })
      }
    }
  }
}

/**
 * Decorator for instrumenting methods with both OpenTelemetry traces and metrics
 *
 * This decorator composes @WithSpan and @WithMetric to provide both trace and metric collection.
 * You can selectively enable/disable traces or metrics using the config options.
 *
 * @param config - Configuration object or string name for the instrumentation
 */
export function WithInstrumentation(config?: string | InstrumentationConfig) {
  const cfg: InstrumentationConfig = typeof config === 'string' ? { name: config } : config || {}
  const { enableTraces = true, enableMetrics = true, name, description, labels } = cfg

  return function <This, Args extends any[], Return>(
    originalMethod: AsyncMethod<This, Args, Return>,
    context: ClassMethodDecoratorContext<This, AsyncMethod<This, Args, Return>>,
  ): AsyncMethod<This, Args, Return> {
    let wrapped: AsyncMethod<This, Args, Return> = originalMethod
    if (enableTraces) {
      wrapped = WithSpan({ name, attributes: labels })(wrapped, context)
    }
    if (enableMetrics) {
      wrapped = WithMetric({ name, description, labels })(wrapped, context)
    }
    return wrapped
  }
}

File: libs/sdk-python/src/daytona/_utils/otel_decorator.py
# Copyright 2025 Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

"""OpenTelemetry instrumentation decorators for tracing and metrics."""

import functools
import time
from typing import Any, Callable, Dict, Optional, TypeVar, cast

from opentelemetry import metrics, trace
from opentelemetry.trace import Status, StatusCode

# Type variable for generic function decoration
F = TypeVar("F", bound=Callable[..., Any])

# Lazy initialization to ensure SDK is started before getting tracer/meter
_tracer = None
_meter = None
_execution_histograms: Dict[str, Any] = {}


def get_tracer():
    """Get or create the tracer instance."""
    global _tracer  # pylint: disable=global-statement
    if _tracer is None:
        _tracer = trace.get_tracer("")
    return _tracer


def get_meter():
    """Get or create the meter instance."""
    global _meter  # pylint: disable=global-statement
    if _meter is None:
        _meter = metrics.get_meter("")
    return _meter


def to_snake_case(string: str) -> str:
    """Converts a string to snake_case for Prometheus-friendly metric names."""
    result = ""
    for i, char in enumerate(string):
        if char.isupper() and i > 0:
            result += "_"
        result += char.lower()
    return result.replace(".", "_")


def with_span(
    name: Optional[str] = None,
    attributes: Optional[Dict[str, str]] = None,
) -> Callable[[F], F]:
    """Decorator for instrumenting methods with OpenTelemetry spans (traces only).

    Args:
        name: Custom name for the span. If not provided, uses `ClassName.methodName` format
        attributes: Additional attributes to attach to the span

    Returns:
        Decorated function with span instrumentation

    Example:
        ```python
        @with_span(name="custom_operation", attributes={"custom": "value"})
        async def my_method(self):
            pass
        ```
    """
    attrs = attributes or {}

    def decorator(func: F) -> F:
        @functools.wraps(func)
        async def wrapper(*args, **kwargs):
            # Get class name if this is a method
            class_name = args[0].__class__.__name__ if args and hasattr(args[0], "__class__") else ""
            method_name = func.__name__

            span_name = name or f"{class_name}.{method_name}" if class_name else method_name

            all_attributes = {
                "component": class_name,
                "method": method_name,
                **attrs,
            }

            tracer = get_tracer()
            with tracer.start_as_current_span(span_name, attributes=all_attributes) as span:
                try:
                    result = await func(*args, **kwargs)
                    span.set_status(Status(StatusCode.OK))
                    return result
                except Exception as error:
                    span.set_status(Status(StatusCode.ERROR, str(error)))
                    span.record_exception(error)
                    raise

        return cast(F, wrapper)

    return decorator


def with_metric(
    name: Optional[str] = None,
    description: Optional[str] = None,
    labels: Optional[Dict[str, str]] = None,
) -> Callable[[F], F]:
    """Decorator for instrumenting methods with OpenTelemetry metrics (metrics only).

    Collects histogram metric:
    - Histogram: `{name}_duration` - tracks execution duration in milliseconds

    Args:
        name: Custom name for the metric. If not provided, uses `ClassName.methodName` format
        description: Description for the metrics being collected
        labels: Additional labels to attach to the metrics

    Returns:
        Decorated function with metric instrumentation

    Example:
        ```python
        @with_metric(name="custom_operation", description="Custom operation duration")
        async def my_method(self):
            pass
        ```
    """
    metric_labels = labels or {}

    def decorator(func: F) -> F:
        @functools.wraps(func)
        async def wrapper(*args, **kwargs):
            # Get class name if this is a method
            class_name = args[0].__class__.__name__ if args and hasattr(args[0], "__class__") else ""
            method_name = func.__name__

            metric_name = to_snake_case(name or f"{class_name}.{method_name}" if class_name else method_name)

            all_labels = {
                "component": class_name,
                "method": method_name,
                **metric_labels,
            }

            # Get or create histogram for this method
            if metric_name not in _execution_histograms:
                meter = get_meter()
                _execution_histograms[metric_name] = meter.create_histogram(
                    f"{metric_name}_duration",
                    description=description or f"Duration of executions for {metric_name}",
                    unit="ms",
                )

            histogram = _execution_histograms[metric_name]
            start_time = time.time()
            status = "success"

            try:
                result = await func(*args, **kwargs)
                return result
            except Exception:
                status = "error"
                raise
            finally:
                duration = (time.time() - start_time) * 1000  # Convert to milliseconds
                histogram.record(duration, {**all_labels, "status": status})

        return cast(F, wrapper)

    return decorator


def with_instrumentation(
    name: Optional[str] = None,
    description: Optional[str] = None,
    labels: Optional[Dict[str, str]] = None,
    enable_traces: bool = True,
    enable_metrics: bool = True,
) -> Callable[[F], F]:
    """Decorator for instrumenting methods with both OpenTelemetry traces and metrics.

    This decorator composes @with_span and @with_metric to provide both trace and metric collection.
    You can selectively enable/disable traces or metrics using the config options.

    Args:
        name: Custom name for the instrumentation
        description: Description for the metrics being collected
        labels: Additional labels/attributes to attach to spans and metrics
        enable_traces: Enable trace collection (default: True)
        enable_metrics: Enable metrics collection (default: True)

    Returns:
        Decorated function with both span and metric instrumentation

    Example:
        ```python
        @with_instrumentation(name="create_sandbox", enable_metrics=True)
        async def create(self, params):
            pass
        ```
    """

    def decorator(func: F) -> F:
        decorated = func

        if enable_metrics:
            decorated = with_metric(name=name, description=description, labels=labels)(decorated)

        if enable_traces:
            decorated = with_span(name=name, attributes=labels)(decorated)

        return decorated

    return decorator
